solr/core/src/test/org/apache/solr/search/RankQueryTest.java

```diff
@@ -106,4 +119,65 @@ public void testPluggableCollector() {
         "//result/doc[2]/str[@name='id'][.='6']",
         "//result/doc[1]/str[@name='id'][.='5']");
   }
+
+  // The following static classes are intended to ensure that support of covariant
+  // ScoreDocs is supported in rank queries. MyRankQuery will fail to compile
+  // if covariant ScoreDocs are not supported because it returns a TopDocsCollector
+  // for MyScoreDoc (a subtype of ScoreDoc).
+  static class MyScoreDoc extends ScoreDoc {
+    public int someOtherField = 0;
+    public MyScoreDoc(int doc, float score, int someOtherField) {
+      super(doc, score, -1);
```

Comment:
minor/subjective

```suggestion
    public MyScoreDoc(int doc, float score, int shardIndex, int someOtherField) {
      super(doc, score, shardIndex);
```

or

```suggestion
    public MyScoreDoc(int doc, float score, int someOtherField) {
      super(doc, score);
```